File: Wayfork/App/Services/LogCenter.swift
import Foundation
import Observation
import WayforkCore
import os

/// In-memory ring of log lines behind the Logs window plus the on-disk mirrors
/// `wayfork.log` / `runtime.log` (docs/design/06-logging.md).
@MainActor
@Observable
final class LogCenter {
    nonisolated static let ringCapacity = 10_000
    nonisolated static let appSource = "app"
    nonisolated static let runtimeFileName = "runtime"
    nonisolated static let appFileName = "wayfork"

    nonisolated static var directory: URL {
        let base =
            FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).first
            ?? FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent("Library")
        return base.appendingPathComponent("Logs/Wayfork", isDirectory: true)
    }

    /// Oldest first, at most `ringCapacity` lines (trimmed in chunks).
    private(set) var lines: [LogLine] = []
    /// Lines below this level are neither stored nor shown.
    var minimumLevel: LogLevel = .info

    private let runtimeFile: AppLogFile?
    private let appFile: AppLogFile?
    private let osLog = Logger(subsystem: WayforkIdentifiers.app, category: "app")
    /// Keys of recently received daemon lines; `subscribe` replays ring buffers and the
    /// tail loaded from `runtime.log` overlaps with them.
    private var seen: Set<LineKey> = []

    private struct LineKey: Hashable {
        let ts: Date
        let source: String
        let message: String

        init(_ line: LogLine) {
            ts = line.ts
            source = line.source
            message = line.message
        }
    }

    init() {
        let directory = LogCenter.directory
        runtimeFile = try? AppLogFile(directory: directory, name: LogCenter.runtimeFileName)
        appFile = try? AppLogFile(directory: directory, name: LogCenter.appFileName)
        if let runtimeFile {
            let tail = runtimeFile.tail(maxLines: LogCenter.ringCapacity)
            lines = tail.compactMap(LogLineFormat.parse)
            seen = Set(lines.suffix(5000).map(LineKey.init))
        }
    }

    // MARK: - App log

    func app(_ level: LogLevel, _ message: String) {
        switch level {
        case .error: osLog.error("\(message, privacy: .public)")
        case .warning: osLog.warning("\(message, privacy: .public)")
        case .info: osLog.info("\(message, privacy: .public)")
        case .debug: osLog.debug("\(message, privacy: .public)")
        }
        let line = LogLine(source: LogCenter.appSource, level: level, message: message)
        guard level <= minimumLevel else { return }
        appFile?.append(LogLineFormat.format(line))
        append([line])
    }

    // MARK: - Daemon stream

    func receive(_ batch: [LogLine]) {
        var fresh: [LogLine] = []
        fresh.reserveCapacity(batch.count)
        for line in batch where line.level <= minimumLevel {
            let key = LineKey(line)
            guard !seen.contains(key) else { continue }
            seen.insert(key)
            fresh.append(line)
            runtimeFile?.append(LogLineFormat.format(line))
        }
        if seen.count > 20_000 {
            seen = Set(lines.suffix(5000).map(LineKey.init))
        }
        append(fresh)
    }

    func clear() {
        lines.removeAll()
    }

    /// Deletes rotated files older than the retention period.
    func prune(retentionDays: Int) {
        let directory = LogCenter.directory
        for name in [LogCenter.runtimeFileName, LogCenter.appFileName] {
            AppLogFile.prune(directory: directory, name: name, retentionDays: retentionDays)
        }
    }

    private func append(_ fresh: [LogLine]) {
        guard !fresh.isEmpty else { return }
        lines.append(contentsOf: fresh)
        let overflow = lines.count - LogCenter.ringCapacity
        if overflow > 1000 {
            lines.removeFirst(overflow)
        }
    }
}

File: WayforkWindows/app/lib/app/services/log_center.dart
import 'dart:collection';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:wayfork/core/app/log_file.dart';
import 'package:wayfork/core/ipc/payloads.dart';
import 'package:wayfork/core/model/settings.dart';
import 'package:wayfork/core/store/store_repository.dart';

/// In-memory ring of log lines behind the Logs page plus the on-disk mirrors
/// `wayfork.log` / `runtime.log` under `%LOCALAPPDATA%\Wayfork\logs`
/// (docs/design/06-logging.md; docs/design/08-windows.md, "Filesystem
/// layout"). Without a directory it keeps the ring only (tests).
final class LogCenter extends ChangeNotifier {
  LogCenter({this.directory, this.echoToConsole = kDebugMode}) {
    final directory = this.directory;
    if (directory == null) return;
    _runtimeFile = _open(directory, runtimeFileName);
    _appFile = _open(directory, appFileName);
    final tail = _runtimeFile?.tail(ringCapacity) ?? const <String>[];
    _lines.addAll(tail.map(LogLineFormat.parse).whereType<LogLine>());
    _seen.addAll(_recentKeys());
  }

  static const ringCapacity = 10000;
  static const appSource = 'app';
  static const runtimeFileName = 'runtime';
  static const appFileName = 'wayfork';

  /// `%LOCALAPPDATA%\Wayfork\logs`.
  static Directory defaultDirectory() {
    final base = StoreRepository.defaultDirectory().path;
    final separator = Platform.pathSeparator;
    return Directory(
      base.endsWith(separator) ? '${base}logs' : '$base${separator}logs',
    );
  }

  final Directory? directory;

  /// Also `debugPrint` every app line (the debug console).
  final bool echoToConsole;

  final _lines = <LogLine>[];
  AppLogFile? _runtimeFile;
  AppLogFile? _appFile;
  LogLevel _minimumLevel = LogLevel.info;

  /// Keys of recently received service lines: `subscribe` replays the
  /// service's ring buffers, and the tail loaded from `runtime.log` overlaps
  /// with them.
  final _seen = <_LineKey>{};

  /// Oldest first, at most [ringCapacity] lines (trimmed in chunks).
  List<LogLine> get lines => UnmodifiableListView(_lines);

  /// Lines below this level are neither stored nor shown.
  LogLevel get minimumLevel => _minimumLevel;
  set minimumLevel(LogLevel level) {
    if (level == _minimumLevel) return;
    _minimumLevel = level;
    notifyListeners();
  }

  // App log

  void app(LogLevel level, String message) {
    if (echoToConsole) debugPrint('[wayfork ${level.jsonValue}] $message');
    final line = LogLine(
      ts: DateTime.now().toUtc(),
      source: appSource,
      level: level,
      message: message,
    );
    if (!_accepts(level)) return;
    _appFile?.append(LogLineFormat.format(line));
    _append([line]);
  }

  // Service stream

  void receive(List<LogLine> batch) {
    final fresh = <LogLine>[];
    for (final line in batch) {
      if (!_accepts(line.level)) continue;
      final key = _LineKey(line);
      if (!_seen.add(key)) continue;
      fresh.add(line);
      _runtimeFile?.append(LogLineFormat.format(line));
    }
    if (_seen.length > 20000) {
      _seen
        ..clear()
        ..addAll(_recentKeys());
    }
    _append(fresh);
  }

  void clear() {
    if (_lines.isEmpty) return;
    _lines.clear();
    notifyListeners();
  }

  /// Deletes rotated files older than the retention period.
  void prune({required int retentionDays}) {
    final directory = this.directory;
    if (directory == null) return;
    for (final name in const [runtimeFileName, appFileName]) {
      AppLogFile.prune(directory, name, retentionDays: retentionDays);
    }
  }

  void close() {
    _runtimeFile?.close();
    _appFile?.close();
  }

  @override
  void dispose() {
    close();
    super.dispose();
  }

  bool _accepts(LogLevel level) => level.rank <= _minimumLevel.rank;

  void _append(List<LogLine> fresh) {
    if (fresh.isEmpty) return;
    _lines.addAll(fresh);
    final overflow = _lines.length - ringCapacity;
    if (overflow > 1000) _lines.removeRange(0, overflow);
    notifyListeners();
  }

  Iterable<_LineKey> _recentKeys() => _lines
      .skip(_lines.length > 5000 ? _lines.length - 5000 : 0)
      .map(_LineKey.new);

  static AppLogFile? _open(Directory directory, String name) {
    try {
      return AppLogFile(directory, name);
    } on AppLogFileException {
      return null;
    }
  }
}

final class _LineKey {
  _LineKey(LogLine line)
    : ts = line.ts,
      source = line.source,
      message = line.message;

  final DateTime ts;
  final String source;
  final String message;

  @override
  bool operator ==(Object other) =>
      other is _LineKey &&
      ts == other.ts &&
      source == other.source &&
      message == other.message;

  @override
  int get hashCode => Object.hash(ts, source, message);
}
